module/zfs/zio.c

```diff
@@ -3105,52 +3163,126 @@ zio_write_gang_block(zio_t *pio, metaslab_class_t *mc)
-		zp.zp_type = zp.zp_storage_type = DMU_OT_NONE;
-		zp.zp_level = 0;
-		zp.zp_copies = gio->io_prop.zp_copies;
-		zp.zp_dedup = B_FALSE;
-		zp.zp_dedup_verify = B_FALSE;
-		zp.zp_nopwrite = B_FALSE;
-		zp.zp_encrypt = gio->io_prop.zp_encrypt;
-		zp.zp_byteorder = gio->io_prop.zp_byteorder;
-		zp.zp_direct_write = B_FALSE;
-		memset(zp.zp_salt, 0, ZIO_DATA_SALT_LEN);
-		memset(zp.zp_iv, 0, ZIO_DATA_IV_LEN);
-		memset(zp.zp_mac, 0, ZIO_DATA_MAC_LEN);
-
-		zio_t *cio = zio_write(zio, spa, txg, &gbh->zg_blkptr[g],
-		    has_data ? abd_get_offset(pio->io_abd, pio->io_size -
-		    resid) : NULL, lsize, lsize, &zp,
-		    zio_write_gang_member_ready, NULL,
+	int g;
+	flags &= METASLAB_ASYNC_ALLOC;
+	flags |= METASLAB_GANG_CHILD;
+	zp.zp_checksum = gio->io_prop.zp_checksum;
+	zp.zp_compress = ZIO_COMPRESS_OFF;
+	zp.zp_complevel = gio->io_prop.zp_complevel;
+	zp.zp_type = zp.zp_storage_type = DMU_OT_NONE;
+	zp.zp_level = 0;
+	zp.zp_copies = gio->io_prop.zp_copies;
+	zp.zp_dedup = B_FALSE;
+	zp.zp_dedup_verify = B_FALSE;
+	zp.zp_nopwrite = B_FALSE;
+	zp.zp_encrypt = gio->io_prop.zp_encrypt;
+	zp.zp_byteorder = gio->io_prop.zp_byteorder;
+	zp.zp_direct_write = B_FALSE;
+	memset(zp.zp_salt, 0, ZIO_DATA_SALT_LEN);
+	memset(zp.zp_iv, 0, ZIO_DATA_IV_LEN);
+	memset(zp.zp_mac, 0, ZIO_DATA_MAC_LEN);
+
+	for (g = 0; resid != 0; g++) {
+		uint64_t min_size = zio_roundup_alloc_size(spa,
+		    resid / (gbh_nblkptrs(gangblocksize) - g));
+		min_size = MIN(min_size, resid);
+		IMPLY(min_size < spa->spa_min_alloc, min_size == resid);
+		IMPLY(min_size >= spa->spa_min_alloc, min_size <= resid);
+		bp = &((blkptr_t *)gbh)[g];
+
+		zio_alloc_list_t cio_list;
+		metaslab_trace_init(&cio_list);
+		error = metaslab_alloc_range(spa, mc, min_size, resid,
+		    bp, gio->io_prop.zp_copies, txg, NULL,
+		    flags, &cio_list, NULL, zio->io_allocator);
+
+		uint64_t allocated_size = 0;
+		for (int d = 0; d < BP_GET_NDVAS(bp); d++) {
+			uint64_t asize = DVA_GET_ASIZE(&bp->blk_dva[d]);
+			if (asize > allocated_size)
+				allocated_size = asize;
+		}
+		boolean_t allocated = allocated_size != 0;
+		if (g == 0 && error == 0 && allocated_size == pio->io_size) {
+			ASSERT3U(BP_GET_NDVAS(bp), ==, gio->io_prop.zp_copies);
+			/*
+			 * De-gang case: We got an allocation big enough to
+			 * satisfy the original allocation. Just do that
+			 * instead of ganging.
+			 */
+			for (int d = 0; d < BP_GET_NDVAS(zio->io_bp); d++) {
+				dva_t *dva = &zio->io_bp->blk_dva[d];
+				metaslab_unalloc_dva(spa,
+				    dva, txg);
+				metaslab_group_alloc_decrement(spa,
+				    DVA_GET_VDEV(dva), pio, flags,
+				    pio->io_allocator, B_FALSE);
+			}
+			metaslab_trace_move(&cio_list, &pio->io_alloc_list);
+			metaslab_group_alloc_increment_all(spa, bp, pio, flags,
+			    pio->io_allocator);
+			if (pio->io_flags & ZIO_FLAG_IO_ALLOCATING) {
+				metaslab_class_throttle_unreserve(mc,
+				    gbh_copies - copies, pio->io_allocator,
+				    pio);
+			}
```

Comment:
In my https://github.com/openzfs/zfs/pull/17020 I am removing this allocator/throttling accounting. We could merge that first before complicating this.